src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -0,0 +1,377 @@
+  const gz::sim::Entity &_entity,
+  const std::shared_ptr<const sdf::Element> &_sdf,
+  gz::sim::EntityComponentManager &/*_ecm*/,
+  gz::sim::EventManager &/*_eventMgr*/)
+{
+  dataPtr->worldEntity = _entity;
+}
+
+////////////////////////////////////////////////////////////////
+void EnvironmentalSensorSystem::PreUpdate(const gz::sim::UpdateInfo &_info,
+  gz::sim::EntityComponentManager &_ecm)
+{
+
+  _ecm.EachNew<components::CustomSensor, components::ParentEntity>(
+    [&](const Entity &_entity,
+        const components::CustomSensor *_custom,
+        const components::ParentEntity *_parent)->bool
+      {
+        // Get sensor's scoped name without the world
+        auto sensorScopedName =
+          removeParentScope(scopedName(_entity, _ecm, "::", false), "::");
+        sdf::Sensor data = _custom->Data();
+        data.SetName(sensorScopedName);
+
+        auto type = gz::sensors::customType(data);
+        if (!common::StartsWith(type, SENSOR_TYPE_PREFIX))
+        {
+          return true;
+        }
+
+        // Default to scoped name as topic
+        if (data.Topic().empty())
+        {
+          std::string topic = scopedName(_entity, _ecm) + "/" + type;
+          data.SetTopic(topic);
+        }
+
+        gz::sensors::SensorFactory sensorFactory;
+        auto sensor = sensorFactory.CreateSensor<EnvironmentalSensor>(data);
+
+        // Set sensor parent
+        auto parentName = _ecm.Component<components::Name>(
+            _parent->Data())->Data();
+        sensor->SetParent(parentName);
+
+        // Set topic on Gazebo
+        _ecm.CreateComponent(_entity,
+            components::SensorTopic(sensor->Topic()));
+
+        // Get current EnvironmentalData component
+        auto environData =
+          _ecm.Component<components::Environment>(
+            worldEntity(_ecm));
+
+        if (environData != nullptr)
+        {
+          sensor->SetDataTable(environData, _info.simTime);
+        }
+        else
+        {
+          gzerr << "No sensor data loaded\n";
+        }
+
+        // Keep track of this sensor
+        this->dataPtr->entitySensorMap.insert(std::make_pair(_entity,
+            std::move(sensor)));
+        return true;
+      });
+}
+
+////////////////////////////////////////////////////////////////
+void EnvironmentalSensorSystem::PostUpdate(const gz::sim::UpdateInfo &_info,
+    const gz::sim::EntityComponentManager &_ecm)
+{
+  _ecm.EachNew<components::Environment>([&](const Entity &/*_entity*/,
+    const components::Environment *_environmental_data)->bool
+    {
+      for (auto &[entity, sensor] : this->dataPtr->entitySensorMap)
+      {
+        sensor->SetDataTable(_environmental_data, _info.simTime);
+      }
+      return true;
+    });
+  // Only update and publish if not paused.
+  if (!_info.paused)
+  {
+
+    for (auto &[entity, sensor] : this->dataPtr->entitySensorMap)
+    {
+      sensor->UpdatePosition(entity, _ecm);
+      sensor->Update(_info.simTime);
```

Comment:
@arjo129 meta: would it make sense to collapse both methods into one?

Reply:
Not really. Update() is part of `gz::sensors::Sensor` base class. I cannot pass it the ECM to allow it to extract update data, hence I would still need `UpdatePosition` separately comewhere.